semgrep-core/src/core/ast/Visitor_AST.ml

```diff
@@ -1233,12 +1234,54 @@ let ii_of_any any =
 
 (*e: function [[Lib_AST.ii_of_any]] *)
 
+(*****************************************************************************)
+(* Extract ranges *)
+(*****************************************************************************)
+
+(*s: function [[Lib_AST.extract_info_visitor]] *)
+let extract_ranges recursor =
+  let ranges = ref None in
+  let smaller t1 t2 =
+    if compare t1.PI.charpos t2.PI.charpos < 0 then t1 else t2
+  in
+  let larger t1 t2 =
+    if compare t1.PI.charpos t2.PI.charpos > 0 then t1 else t2
+  in
+  let incorporate_tokens (left, right) =
+    match !ranges with
+    | None -> ranges := Some (left, right)
+    | Some (orig_left, orig_right) ->
+        ranges := Some (smaller orig_left left, larger orig_right right)
+  in
+  let incorporate_token tok =
+    if PI.is_origintok tok then
+      let tok_loc = PI.token_location_of_info tok in
+      incorporate_tokens (tok_loc, tok_loc)
+  in
+  let hooks =
+    {
+      default_visitor with
+      kinfo = (fun (_k, _) i -> incorporate_token i);
+      kstmt =
+        (fun (k, _) stmt ->
+          match stmt.s_range with
+          | None -> (
+              let saved_ranges = !ranges in
+              ranges := None;
+              k stmt;
+              stmt.s_range <- !ranges;
+              match saved_ranges with
+              | None -> ()
+              | Some r -> incorporate_tokens r )
+          | Some range -> incorporate_tokens range);
+    }
+  in
+  let vout = mk_visitor hooks in
+  recursor vout;
+  match !ranges with Some range -> range | None -> failwith "no tokens found"
+
 let range_of_tokens tokens =
-  List.filter Parse_info.is_origintok tokens |> Parse_info.min_max_ii_by_pos
+  List.filter PI.is_origintok tokens |> PI.min_max_ii_by_pos
   [@@profiling]
 
-let range_of_any any =
-  let leftmost_token, rightmost_token = ii_of_any any |> range_of_tokens in
-  ( Parse_info.token_location_of_info leftmost_token,
-    Parse_info.token_location_of_info rightmost_token )
-  [@@profiling]
+let range_of_any any = extract_ranges (fun visitor -> visitor any) [@@profiling]
```

Comment:
maybe you could add in the .mli that this function will have some side effects and set the s_range fields for any substmt in any.